From d8d2d6490f76cd2a34a382b50e43b29db83e26a8 Mon Sep 17 00:00:00 2001
From: Ray Chang <ray.chang@technexion.com>
Date: Mon, 4 Aug 2025 09:45:18 +0800
Subject: [PATCH] xtoptions: add error checking for memory

Using MASQUERADE target call xtables_options_xfrm() causes memory
issues.

Upstream-Status: Pending
---
 libxtables/xtoptions.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/libxtables/xtoptions.c b/libxtables/xtoptions.c
index 64d6599..ccad2a5 100644
--- a/libxtables/xtoptions.c
+++ b/libxtables/xtoptions.c
@@ -88,12 +88,21 @@ xtables_options_xfrm(struct option *orig_opts, struct option *oldopts,
 		     const struct xt_option_entry *entry, unsigned int *offset)
 {
 	int num_new, i;
-	struct option *merge, *mp;
+	struct option *merge, *mp = NULL;
+
+	/* Validate input parameters */
+	if (!entry) {
+		return NULL;
+	}

 	for (num_new = 0; entry[num_new].name != NULL; ++num_new)
 		;

 	mp = xtables_calloc(num_new + 1, sizeof(*mp));
+	if (!mp) {
+		return NULL;
+	}
+
 	for (i = 0; i < num_new; i++) {
 		mp[i].name	= entry[i].name;
 		mp[i].has_arg	= entry[i].type != XTTYPE_NONE;
--
2.50.1

